File: easy/valid-parentheses.py
def isValid(s):
    open_order = list()

    for i in s:
        if i in ['(','{','[']:
            open_order.append(i)
        elif i in [')', '}', ']']:
            if len(open_order) == 0:
                return False
            match i:
                case ')':
                    if open_order.pop() != '(':
                        return False
                case '}':
                    if open_order.pop() != '{':
                        return False
                case ']':
                    if open_order.pop() != '[':
                        return False
    if open_order:
        return False
    else:
        return True
                    
import unittest

class TestIsValid(unittest.TestCase):

    def test_first_sample(self):
        self.assertEqual(isValid("()"), True)

    def test_second_sample(self):
        self.assertEqual(isValid("()[]{}"), True)

    def test_third_sample(self):
        self.assertEqual(isValid("(]"), False)

if __name__ == '__main__':
    unittest.main()

File: easy/valid-parentheses.cpp
#include <iostream>
#include <stack>
#include <string>
#include <cassert>

bool isValid(std::string s) {
    std::stack<char> open_order;

    for (char& i : s) {
        if (i == '(' || i == '{' || i == '[') {
            open_order.push(i);
        } else if (i == ')' || i == '}' || i == ']') {
            if (open_order.empty()) {
                return false;
            }
            char match = open_order.top();
            open_order.pop();
            if ((i == ')' && match != '(') ||
                (i == '}' && match != '{') ||
                (i == ']' && match != '[')) {
                return false;
            }
        }
    }

    return open_order.empty();
}

int main() {
    // Test 1
    assert(isValid("()") == true);

    // Test 2
    assert(isValid("()[]{}") == true);

    // Test 3
    assert(isValid("(]") == false);

    std::cout << "All tests passed successfully.\n";
    return 0;
}

